companion/src/modeledit/curves.cpp

```diff
@@ -588,41 +565,68 @@ void Curves::on_curveApply_clicked()
   emit modified();
 }
 
-#if 0
-void ModelEdit::clearCurves(bool ask)
+void Curves::ShowContextMenu(const QPoint& pos) // this is a slot
 {
-    if (ask) {
-      int res = QMessageBox::question(this, tr("Clear Curves?"), tr("Really clear all the curves?"), QMessageBox::Yes | QMessageBox::No);
-      if (res!=QMessageBox::Yes) return;
-    }
-    curvesLock=true;
-    for (int j=0; j<16; j++) {
-      model->curves[j].count = 5;
-      model->curves[j].custom = false;
-      memset(model->curves[j].name, 0, sizeof(model->curves[j].name));
-      for (int i=0; i<17; i++) {
-        model->curves[j].points[i].x = 0;
-        model->curves[j].points[i].y = 0;
+  QPushButton *button = (QPushButton *)sender();
+  int index = button->property("index").toInt();
+  
+  const QClipboard *clipboard = QApplication::clipboard();
+  const QMimeData *mimeData = clipboard->mimeData();
+  QWidget * mywidget=(QWidget *)sender();
+  QPoint globalPos = mywidget->mapToGlobal(pos);
+
+  QMenu myMenu;
+  QAction *action;
+  action = myMenu.addAction(CompanionIcon("copy.png"),tr("Copy"));
+  action->setProperty("index",CURVE_COPY);
+  
+  action = myMenu.addAction(CompanionIcon("paste.png"),tr("Paste"));
+  if (!mimeData->hasFormat("application/x-companion-curve-item")) {
+    action->setEnabled(false);
+  }
+  action->setProperty("index",CURVE_PASTE);
+  
+  action = myMenu.addAction(CompanionIcon("clear.png"),tr("Clear"));
+  action->setProperty("index",CURVE_RESET);
+  action = myMenu.addAction(CompanionIcon("clear.png"),tr("Clear all curves"));
+  action->setProperty("index",CURVE_RESETALL);
+
+  QAction* selectedItem = myMenu.exec(globalPos);
+  if (selectedItem)
+  {
+    int action=selectedItem->property("index").toInt();
+    if (action==CURVE_COPY) {
+      QByteArray curveData;
+      QMimeData *mimeData2 = new QMimeData;
+      curveData.append((char*)&model->curves[index], sizeof(CurveData));
+      mimeData2->setData("application/x-companion-curve-item", curveData);
+      QApplication::clipboard()->setMimeData(mimeData2, QClipboard::Clipboard);
+    } 
+    else if (action==CURVE_PASTE) {
+      QByteArray curveData = mimeData->data("application/x-companion-curve-item");
+      CurveData *curve = &model->curves[index];
+      memcpy(curve, curveData.mid(0, sizeof(CurveData)).constData(), sizeof(CurveData));
```

Comment:
So ... curveData is a QByteArray, if we call mid, it constructs a new QByteArray starting at position 0! Why don't we just call curveData.constData(), it would do the same thing, no?